python/ray/data/_internal/logical/rules/operator_fusion.py

```diff
@@ -63,9 +65,14 @@ def _can_fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator) -> bool:
         down_logical_op = self._op_map[down_op]
         up_logical_op = self._op_map[up_op]
 
-        # We only support fusing upstream reads and maps with downstream maps.
-        if not isinstance(down_logical_op, AbstractUDFMap) or not isinstance(
-            up_logical_op, (Read, AbstractUDFMap)
+        # Having Read as downstream or Write as upstream is invalid: Read can only be
+        # the first operator and Write can only be the last operator.
+        if isinstance(down_logical_op, Read) or isinstance(up_logical_op, Write):
+            return False
```

Comment:
We should raise ValueError then? Otherwise, it would be better if the fusion rule didn't inspect the types this closely.

Reply:
This method is not for validating the operator DAG and is just for querying if two operators can be fused, so it seems better not to raise error.

We can remove this check, if: 1) operator DAG is valid (ensured by the caller); 2) this method is called for two operators along the edge of the DAG. We may assume 1 here, but it's probably still good to guard against the 2.

 Do you think we'd better not check Read/Write operators here? They are logical operators though, so seems fine to check them.

Reply:
It seems weird to explicitly check / disallow then to be fused. There's no reason we cannot fuse these in principle if they truly are Map subtypes.

Reply:
Changed it to check whether the downstream op takes input (for Read).
For Write, it can in theory to have a Write->MapBatches fusion where the MapBatches will just process the WriteResults, so drop the check for Write here.